pkg/controller/cluster_pools_pull_secret_provider/cluster_pools_pull_secret_provider.go

```diff
@@ -0,0 +1,142 @@
+package cluster_pools_pull_secret_provider
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+
+	"github.com/openshift/ci-tools/pkg/api"
+	controllerutil "github.com/openshift/ci-tools/pkg/controller/util"
+	"github.com/openshift/ci-tools/pkg/util"
+)
+
+const ControllerName = "cluster_pools_pull_secret_provider"
+
+func AddToManager(hiveManager manager.Manager,
+	sourcePullSecretNamespace, sourcePullSecretName string) error {
+	log := logrus.WithField("controller", ControllerName)
+
+	r := &reconciler{
+		log:                       log,
+		client:                    hiveManager.GetClient(),
+		sourcePullSecretNamespace: sourcePullSecretNamespace,
+		sourcePullSecretName:      sourcePullSecretName,
+	}
+	c, err := controller.New(ControllerName, hiveManager, controller.Options{
+		Reconciler: r,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to construct controller: %w", err)
+	}
+
+	if err := c.Watch(
+		source.NewKindWithCache(&hivev1.ClusterPool{}, hiveManager.GetCache()),
+		clusterPoolHandler(sourcePullSecretNamespace),
+	); err != nil {
+		return fmt.Errorf("failed to create watch for testimagestreamtagimports: %w", err)
+	}
+
+	r.log.Info("Successfully added reconciler to manager")
+	return nil
+}
+
+func clusterPoolHandler(sourcePullSecretNamespace string) handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(o ctrlruntimeclient.Object) []reconcile.Request {
+		clusterPool, ok := o.(*hivev1.ClusterPool)
+		if !ok {
+			logrus.WithField("type", fmt.Sprintf("%T", o)).Error("Got object that was not an ClusterPool")
+			return nil
+		}
+		if clusterPool.Namespace == sourcePullSecretNamespace {
+			return nil
+		}
+		return []reconcile.Request{{NamespacedName: types.NamespacedName{
+			Namespace: clusterPool.Namespace,
+			Name:      clusterPool.Name,
+		}}}
+	})
+}
+
+type reconciler struct {
+	log                       *logrus.Entry
+	client                    ctrlruntimeclient.Client
+	sourcePullSecretNamespace string
+	sourcePullSecretName      string
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithField("request", req.String())
+	err := r.reconcile(ctx, req, log)
```

Comment:
```suggestion
	if err := r.reconcile(ctx, req, log); err != nil && !apierrors.IsAlreadyExists(err) {
		log.WithError(err).Error("Reconciliation failed")
	} else {
		log.Info("Finished reconciliation")
	}
```
Also that means that we will log `Finished reconciliation` if the errors is `Already Exists`. Do we care?

Reply:
`err` is used outside the `if` sentence.

Reply:
(And it is extremely important that that doesn't get changed, as it is responsible for the control flow (if err != nil means it will get retried after a backoff))